app/frontend/src/features/search/useSearchFilters.ts

```diff
@@ -0,0 +1,88 @@
+import { Location } from "history";
+import { useCallback, useEffect, useRef, useState } from "react";
+import { useHistory, useLocation } from "react-router-dom";
+import { UserSearchFilters as ServiceUserSearchFilters } from "service/search";
+
+export interface SearchFilters extends ServiceUserSearchFilters {
+  location?: string;
+}
+
+export default function useSearchFilters(route: string) {
+  const location = useLocation();
+  const [active, setActive] = useState<SearchFilters>(
+    locationToFilters(location)
+  );
+  const pending = useRef<SearchFilters>(locationToFilters(location));
+
+  const history = useHistory();
+
+  useEffect(() => {
+    const entries: [string, string][] = [];
+    Object.entries(active).forEach((entry) => {
+      if (typeof entry[1] === "object") {
+        Object.values(entry[1]).forEach((value) => {
+          entries.push([entry[0], `${value}`]);
+        });
+      } else {
+        entries.push(entry);
+      }
+    });
+    const searchParams = new URLSearchParams(entries);
+    history.push(`${route}?${searchParams.toString()}`);
+  }, [active, history, route]);
+
+  const change = useCallback(
+    <T extends keyof SearchFilters>(filter: T, value: SearchFilters[T]) => {
+      pending.current = { ...pending.current, [filter]: value };
+      if (value === undefined) delete pending.current[filter];
+    },
+    []
+  );
+
+  const apply = useCallback(() => {
+    setActive(pending.current);
+  }, []);
+
+  return { active, change, apply };
+}
+
+export function locationToFilters(location: Location) {
+  const searchParams = new URLSearchParams(location.search);
+  const filters: SearchFilters = {};
+  Array.from(searchParams.keys()).forEach((untypedkey) => {
+    const key = untypedkey as unknown as keyof SearchFilters;
```

Comment:
```suggestion
    const key = untypedkey as keyof SearchFilters;
```
Would this alone not work since `untypedkey` would be a string to start with?

Reply:
Weird, I thought I tried this.